Various test fixes, `@php` directive support

claytonrcarter opened the pull request:
Hi there! This does a few things:

1. tweaks `formatAsPhp` to correct what seemed like some some logic mistakes
In particular, `manipulated` was bring created but was not being used for the formatting, so most php code was not being formatted. This change corrects that. OF the changes included here, this one feels the most solid. I think the original code was just wrong.

2. implements a hack-around for some wrapping issues w/ various directives
This is the code in the 2nd and 3rd commits: 7e5c970 and 9a270da. As you know, the plugin works by replacing Blade code w/ html-ish elements so that they can be formatted as HTML, then it replaces those placeholders w/ the formatted PHP. The issue w/ this is that something like `Hi, my name is {{ $name }}` is being formatted as HTML that looks like `Hi, my name is <echo-really-long-uuid-here />`, and if that `<echo...>` element happens to run over 80 chars (or whatever the line legth is), it will be wrapped. Then, when the PHP is replaced back it, we'll ended up w/ 
```
Hi, my name is
{{ $name }}
```

My "fix" is just a hack that trims these placeholders to match the length of the input code. **This is not a final** solution, but maybe it gets closer? The issue w/ this fix is that if the input code change in length when it's formatted as PHP, then it could mean that we haven't wrapped that line correctly. eg `Pretend this line is only {{79}} chars` will get turned into `Pretend this line is only {{ 79 }} chars`, which is actually 81 chars (b/c of the 2 added spaces around `79`). This feels "better" than the above example, but it's obviously not ideal.

3. Adds support for `@php` directives.
This went OK, I think, but feels very, very boilerplatey. The couple of new tests pass.

This was mostly a learning excercise for me; if these changes aren't desirable (or they're too hacky or just wrong!) it was at least fun to learn my way around the project.

olivernybroe commented on src/lang/nodes/index.ts:
```diff
@@ -183,6 +200,25 @@ export class VerbatimNode implements Node {
+
+    toHtml(): HtmlOutput {
+        return {
+            asHtml: `<php-${randomUUID()} />`,
```
You can use the function `nextId` instead of UUID. It's a short number incrementing each time you use it.
If you want to force a newline, we have the const `forceHtmlSplit` that can be added in 👍

claytonrcarter replied:
Thank you. Unless you think that would be better here, I suspect that the UUID version is fine, since I suspect it's very unlikely that a `@php` block will be shorter than than a UUID placeholder. Maybe I'm wrong? I support something like `@php $foo = 'bar' @endphp` would be pretty short...

Is there any reason to use the UUID, then? Or should it just be completely replaced w/ `nextId()`?

olivernybroe replied:
I think we could stop using UUID completely, it was just there because I hadn't created the nextId function 🙂

We achieve the same with it, just shorter and a quicker calculation

olivernybroe commented:
First of all, thank you a lot for this work, it looks really great and def. something we should have into the codebase!  
I am amazed that you found out how to do all this stuff without any documentation being written for it 🤩 

1. Awesome, I have been looking at this part many times and just not getting why it didn't work 😆 

2. Hmm, so I see your issue and that is a real concern. I have a suggestion to what we could do instead though. So I have the function called `nextId` which was an attempt to solve these very long html tags from uuid. It is basically just an integer that increments every time you call it, making the id's unique this way, but much shorter.  
You can then just repeat a character afterwards, so the HTML is the same length as the content. We then have a requirement of a minimum length, so really short echo's will not be 100% correct.  
e.g. `{{ $a }}` which has a length of 8, will be `<echo-13 />`, that has a length of 11.   
A longer example `{{ "My super useless echo" }}`, length 29, will be `<echo-14-xxxxxxxxxxxxxxxxx />`, length 29.  

In regards to your other issue with the php code inside being formatted, resulting in the length to change. What if you just move the `formatAsPhp` call out of `toString` and into `toHtml`, that way you can pass the output of that call into the `trimmedPlaceholderElement` instead of content? 


4. Great job, it looks completely correct and yeah is very boilerplatey to do. I am hoping we can find some time to clean it up and make the process faster to do. It makes sense that you have to manipulate the tokenizer, CST and AST parser and the traverser for printing, but it feels really tedious.

claytonrcarter commented:
Thank you! I just rebased this and pushed some updates, here's my summary:

1. first commit should be the same logically, but I added some (flagged) debugging output to assist w/ debugging of formatting issues. I'm happy to remove these if desired, but I think they'd be helpful as we build this up, and perhaps can be removed later. Your call, just let me know.

2. THANK YOU. That makes total sense, and I've implemented it here. See inline comments FMI.

3. Third commit (`@php` support) hasn't changed.

claytonrcarter commented on src/utils.ts:
```diff
@@ -3,50 +3,94 @@ import { format, ParserOptions } from "prettier";
+
+    // Pad the placeholder to be as long as the "content" so it can be wrapped
+    // correctly.
+    return `<${prefix}-${nextId()}-${"x".repeat(lengthToPad - 1)} />`;
```
The `- 1` in `repeat()` is to account for addition of the `-`. This means that some placeholders may look like `<e-1- />`

olivernybroe reviewed (approved):
This is really nice! 

Let's get rid of the UUID so we only have one approach and then I'll merge this in 🤩

claytonrcarter commented on tests/__fixtures__/comment/basic.blade.php:
```diff
@@ -1,6 +1,4 @@
-<h1>
-    Here's some very nice text {{-- This is actually not that nice --}}
-</h1>
+<h1>Here's some very nice text {{-- This is actually not that nice --}}</h1>
```
@olivernybroe This PR reverts a few changes you made over the weekend in 94ecc89, such as this one. In this case, this is the "correct" formatting according to the prettier that I have installed, and I suspect reflects that new changes in how we pad the placeholders.

claytonrcarter commented on tests/__fixtures__/components/component-directive.blade.php:
```diff
@@ -7,10 +7,10 @@
-    @slot('name', 'value')
-    @slot('name2', true)
-    @slot('name3')
+@component("view")
```
There will be a lot of changes like this. The fixes in `formatAsPhp` mean that all of the fixtures have to be changed from single quotes to double quotes. 

For example, to format the contents of the directive (ie `'view'`) it's turned into `<?php 'view';` which prettier changes to `<?php "view";` because we haven't explicitly told prettier to use single quotes.

claytonrcarter commented on tests/lexer.test.ts:
```diff
@@ -65,7 +65,7 @@ it("should generate echo token when starting with double negation", () => {
 
 it("should generate raw echo tokens when wrapped in parenthesis", () => {
-    const tokens = lex("{{!! 'yes' !!}}").filter(
+    const tokens = lex("({!! 'yes' !!})").filter(
```
I *think* that this test was wrong, but I don't know. It was failing for me and I changed it from `{...}` to `(...)` based on the description of the test. Does this look correct?

claytonrcarter commented on src/utils.ts:
```diff
@@ -3,50 +3,94 @@ import { format, ParserOptions } from "prettier";
+    return formatted;
+};
+
+let id = 1;
+
+export const nextId = () => {
+    return ++id;
+};
```
Note that I also pulled `nextId` into this file b/c it is needed both by `placeholderElement` and in the index file.

claytonrcarter commented:
> Let's get rid of the UUID so we only have one approach and then I'll merge this in 🤩

Done. 🙌  Thank you!

olivernybroe reviewed (approved):
Awesome. let's get this in!
